tools/celtest/test_runner.go

```diff
@@ -0,0 +1,708 @@
+				protocmp.SortRepeatedFields(&exprpb.MapValue{}, "entries")); diff != "" {
+				return TestResult{Success: false, Wanted: fmt.Sprintf("simple value %v", want), Error: fmt.Errorf("mismatched test output with diff (-want +got):\n%s", diff)}
+			}
+			return successResult
+		}, nil
+	case *conformancepb.TestOutput_ResultExpr:
+		return func(val ref.Val, err error) TestResult {
+			if err != nil {
+				return TestResult{Success: false, Error: err}
+			}
+			testOut, err := tr.eval(testOutput.ResultExpr)
+			if err != nil {
+				return TestResult{Success: false, Error: fmt.Errorf("eval(%q) failed: %v", testOutput.ResultExpr, err)}
+			}
+			if optOut, ok := val.(*types.Optional); ok {
+				if optOut.Equal(types.OptionalNone) == types.True {
+					if testOut.Equal(types.OptionalNone) != types.True {
+						return TestResult{Success: false, Wanted: fmt.Sprintf("optional value %v", testOut), Error: fmt.Errorf("policy eval got %v", val)}
+					}
+				} else if testOut.Equal(optOut.GetValue()) != types.True {
+					return TestResult{Success: false, Wanted: fmt.Sprintf("optional value %v", testOut), Error: fmt.Errorf("policy eval got %v", val)}
+				}
+			} else if val.Equal(testOut) != types.True {
+				return TestResult{Success: false, Wanted: fmt.Sprintf("optional value %v", testOut), Error: fmt.Errorf("policy eval got %v", val)}
+			}
+			return successResult
+		}, nil
+	case *conformancepb.TestOutput_EvalError:
+		return func(val ref.Val, err error) TestResult {
+			failureResult := TestResult{Success: false, Wanted: fmt.Sprintf("error %v", testOutput.EvalError)}
+			if err == nil {
+				return failureResult
+			}
+			// Compare the evaluated error with the expected error message only.
+			for _, want := range testOutput.EvalError.GetErrors() {
+				if strings.Contains(err.Error(), want.GetMessage()) {
+					return successResult
+				}
+			}
+			return failureResult
+		}, nil
+	case *conformancepb.TestOutput_Unknown:
+		//  TODO: to implement
+	}
+	return nil, nil
+}
+
+func refValueToExprValue(refVal ref.Val) (*exprpb.ExprValue, error) {
+	if types.IsUnknown(refVal) {
+		return &exprpb.ExprValue{
+			Kind: &exprpb.ExprValue_Unknown{
+				Unknown: &exprpb.UnknownSet{
+					Exprs: refVal.Value().([]int64),
+				},
+			}}, nil
+	}
+	v, err := cel.RefValueToValue(refVal)
+	if err != nil {
+		return nil, err
+	}
+	return &exprpb.ExprValue{
+		Kind: &exprpb.ExprValue_Value{Value: v}}, nil
+}
+
+func canonicalValueToV1Alpha1(val *celpb.Value) (*exprpb.Value, error) {
+	var v1val exprpb.Value
+	b, err := prototext.Marshal(val)
+	if err != nil {
+		return nil, err
+	}
+	if err := prototext.Unmarshal(b, &v1val); err != nil {
+		return nil, err
+	}
+	return &v1val, nil
+}
+
+func (tr *TestRunner) eval(expr string) (ref.Val, error) {
+	e, err := tr.CreateEnv()
+	if err != nil {
+		return nil, err
+	}
+	e, err = e.Extend(cel.OptionalTypes())
+	if err != nil {
+		return nil, fmt.Errorf("e.Extend() failed: %v", err)
+	}
+	ast, iss := e.Compile(expr)
```

Comment:
You can probably wrap the `CreateEnv` and extension with optional types into a `sync.DoOnce` to avoid building it each time